internal/impl/envpath/pathstack.go

```diff
@@ -0,0 +1,99 @@
+package envpath
+
+import (
+	"strings"
+
+	"github.com/samber/lo"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	// PathStackEnv stores the string representation of the stack, as a ":" separated list.
+	// Each element in the list is also the key to the env-var that stores the
+	// nixEnvPath for that devbox-project. Except for the last element, which is InitPathEnv.
+	PathStackEnv = "DEVBOX_PATH_STACK"
+
+	// InitPathEnv stores the path prior to any devbox shellenv modifying the environment
+	InitPathEnv = "DEVBOX_INIT_PATH"
+)
+
+// stack has the following design:
+// 1. The stack enables tracking which sub-paths in PATH come from which devbox-project
+// 2. It is an ordered-list of keys to env-vars that store nixEnvPath values of devbox-projects.
+// 3. The final PATH is reconstructed by concatenating the env-var values of each nixEnvPathKey.
+// 5. The stack is stored in its own env-var PathStackEnv, shared by all devbox-projects in this shell.
+type stack struct {
+
+	// keys holds the stack elements.
+	// Earlier (lower index number) keys get higher priority.
+	// This keeps the string representation of the stack aligned with the PATH value.
+	keys []string
+}
+
+func Stack(env map[string]string) *stack {
+	stackEnv, ok := env[PathStackEnv]
+	if !ok {
+		// if path stack is empty, then push the current PATH, which is the
+		// external environment prior to any devbox-shellenv being applied to it.
+		stackEnv = InitPathEnv
+		env[InitPathEnv] = env["PATH"]
+	}
+	return &stack{
+		keys: strings.Split(stackEnv, ":"),
```

Comment:
Not sure if this matters, but `keys` will be a slice with a single empty string if `stackEnv` is an empty string.

Reply:
good point! I can check for stackEnv being an empty string